rustls/src/server/mod.rs

```diff
@@ -146,6 +146,311 @@ impl<'a> ClientHello<'a> {
+            .with_default_ciphersuites()
+            .with_default_kx_groups()
+    }
+
+    /// Choose a specific set of ciphersuites.
+    pub fn with_ciphersuites(
+        self,
+        ciphersuites: &[&'static SupportedCipherSuite],
+    ) -> ServerConfigBuilderWithSuites {
+        ServerConfigBuilderWithSuites {
+            ciphersuites: ciphersuites.to_vec(),
+        }
+    }
+
+    /// Choose the default set of ciphersuites.
+    ///
+    /// Note that this default provides only high-quality suites: there is no need
+    /// to filter out low-, export- or NULL-strength ciphersuites: rustls does not
+    /// implement these.
+    pub fn with_default_ciphersuites(self) -> ServerConfigBuilderWithSuites {
+        self.with_ciphersuites(DEFAULT_CIPHERSUITES)
+    }
+}
+
+/// A ServerConfigBuilder where we know the ciphersuites.
+pub struct ServerConfigBuilderWithSuites {
+    ciphersuites: Vec<&'static SupportedCipherSuite>,
+}
+
+impl ServerConfigBuilderWithSuites {
+    /// Choose a specific set of key exchange groups.
+    pub fn with_kx_groups(
+        self,
+        kx_groups: &[&'static SupportedKxGroup],
+    ) -> ServerConfigBuilderWithSuitesAndKxGroups {
+        ServerConfigBuilderWithSuitesAndKxGroups {
+            ciphersuites: self.ciphersuites,
+            kx_groups: kx_groups.to_vec(),
+        }
+    }
+
+    /// Choose the default set of key exchange groups.
+    ///
+    /// This is a safe default: rustls doesn't implement any poor-quality groups.
+    pub fn with_default_kx_groups(self) -> ServerConfigBuilderWithSuitesAndKxGroups {
+        self.with_kx_groups(&ALL_KX_GROUPS)
+    }
+}
+
+/// A ServerConfigBuilder where we know the ciphersuites and key exchange
+/// groups.
+pub struct ServerConfigBuilderWithSuitesAndKxGroups {
+    ciphersuites: Vec<&'static SupportedCipherSuite>,
+    kx_groups: Vec<&'static SupportedKxGroup>,
+}
+
+/// Reduce typing for the most common case.
+impl Default for ServerConfigBuilderWithSuitesAndKxGroups {
+    fn default() -> Self {
+        ServerConfigBuilder::new()
+            .with_default_ciphersuites()
+            .with_default_kx_groups()
+    }
+}
+
+impl ServerConfigBuilderWithSuitesAndKxGroups {
+    /// Choose how to verify client certificates.
+    pub fn with_client_cert_verifier(
+        self,
+        client_cert_verifier: Arc<dyn verify::ClientCertVerifier>,
+    ) -> ServerConfigBuilderWithSuitesAndClientAuth {
+        ServerConfigBuilderWithSuitesAndClientAuth {
+            ciphersuites: self.ciphersuites,
+            kx_groups: self.kx_groups,
+            verifier: client_cert_verifier,
+        }
+    }
+
+    /// Disable client authentication.
+    pub fn without_client_auth(self) -> ServerConfigBuilderWithSuitesAndClientAuth {
+        ServerConfigBuilderWithSuitesAndClientAuth {
+            ciphersuites: self.ciphersuites,
+            kx_groups: self.kx_groups,
+            verifier: verify::NoClientAuth::new(),
+        }
+    }
```

Comment:
I'm not sure I would make `without_client_auth()` required. (If you only did that to make the type state work out, there are ways to work around that, although things get a little bit more verbose.)

Reply:
I agree with the idea of making the user explicitly choose between client auth or no client auth. It's a security vulnerability for a server to be configured without client auth unless it is prepared to deal with unauthenticated (at the TLS level) clients.

Reply:
Fair enough.